File: test/a2a-client.test.mjs
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { AgentA2AClient, PublicA2AClient } from "../src/a2a-client.mjs";
import { close, createJSONServer, listen } from "./helpers.mjs";

describe("AgentA2AClient", () => {
  it("calls the runtime call-agent endpoint with current run context", async () => {
    let received = null;
    const server = createJSONServer((req, body) => {
      received = { req, body };
      return { body: { run_id: "child-run-1", status: "running" } };
    });
    const address = await listen(server);
    const apiBase = `http://127.0.0.1:${address.port}`;
    try {
      const client = new AgentA2AClient({ apiBase, runtimeToken: "ol_live_runtime" });
      const result = await client.callAgent({
        currentRunId: "parent-run-1",
        targetAgentId: "target-agent-1",
        input: { q: "search" },
        reason: "delegate",
      });

      assert.equal(result.run_id, "child-run-1");
      assert.equal(received.req.method, "POST");
      assert.equal(received.req.url, "/api/v1/agent-runtime/call-agent");
      assert.equal(received.req.headers.authorization, "Bearer ol_live_runtime");
      assert.deepEqual(received.body, {
        current_run_id: "parent-run-1",
        target_agent_id: "target-agent-1",
        reason: "delegate",
        input: { q: "search" },
      });
    } finally {
      await close(server);
    }
  });
});

describe("PublicA2AClient", () => {
  it("sends JSON-RPC SendMessage with A2A-Version 1.0", async () => {
    let received = null;
    const server = createJSONServer((req, body) => {
      received = { req, body };
      return { body: { jsonrpc: "2.0", id: body.id, result: { id: "task-1", status: { state: "completed" } } } };
    });
    const address = await listen(server);
    const apiBase = `http://127.0.0.1:${address.port}`;
    try {
      const client = new PublicA2AClient({ apiBase, token: "user-token" });
      const result = await client.sendMessage({ slug: "demo-agent", text: "hello", messageId: "msg-1" });

      assert.equal(result.id, "task-1");
      assert.equal(received.req.url, "/api/v1/a2a/agents/demo-agent");
      assert.equal(received.req.headers.authorization, "Bearer user-token");
      assert.equal(received.req.headers["a2a-version"], "1.0");
      assert.equal(received.body.method, "SendMessage");
      assert.equal(received.body.params.message.parts[0].text, "hello");
    } finally {
      await close(server);
    }
  });
});

File: internal/agentnode/a2a_client_test.go
package agentnode

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	openlinker "github.com/OpenLinker-ai/openlinker-go"
)

func TestAgentA2AClientCallAgent(t *testing.T) {
	var received map[string]any
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/agent-runtime/call-agent" {
			t.Fatalf("path = %s", r.URL.Path)
		}
		if got := r.Header.Get("authorization"); got != "Bearer ol_live_test" {
			t.Fatalf("authorization = %q", got)
		}
		if err := json.NewDecoder(r.Body).Decode(&received); err != nil {
			t.Fatal(err)
		}
		writeJSON(w, http.StatusOK, JSONMap{"run_id": "child-run", "status": "success", "output": JSONMap{"answer": "ok"}})
	}))
	defer server.Close()

	client := AgentA2AClient{APIBase: server.URL, RuntimeToken: "ol_live_test"}
	result, err := client.CallAgent(context.Background(), "run-parent", "target-agent", JSONMap{"q": "hello"}, CallAgentOptions{
		Reason: "delegate",
		TaskCallback: &TaskCallbackConfig{
			URL:        "https://caller.example.com/a2a/events",
			Token:      "caller-token",
			Secret:     "caller-secret",
			EventTypes: []string{"run.completed", "run.failed"},
			Metadata:   JSONMap{"client": "agent-node"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if received["current_run_id"] != "run-parent" || received["target_agent_id"] != "target-agent" {
		t.Fatalf("unexpected body: %#v", received)
	}
	push, ok := received["task_callback"].(map[string]any)
	if !ok || push["url"] != "https://caller.example.com/a2a/events" || push["token"] != "caller-token" {
		t.Fatalf("task_callback = %#v", received["task_callback"])
	}
	if push["secret"] != "caller-secret" {
		t.Fatalf("task_callback secret = %#v", push["secret"])
	}
	body := result.(map[string]any)
	if body["run_id"] != "child-run" {
		t.Fatalf("run_id = %v", body["run_id"])
	}
}

func TestAgentA2AClientCallAgentValidationAndErrors(t *testing.T) {
	client := AgentA2AClient{APIBase: "http://127.0.0.1:1", RuntimeToken: "ol_live_test"}
	if _, err := client.CallAgent(context.Background(), "", "target-agent", JSONMap{}, CallAgentOptions{}); err == nil || !strings.Contains(err.Error(), "currentRunID") {
		t.Fatalf("empty currentRunID error = %v", err)
	}
	if _, err := client.CallAgent(context.Background(), "run-parent", "", JSONMap{}, CallAgentOptions{}); err == nil || !strings.Contains(err.Error(), "targetAgentID") {
		t.Fatalf("empty targetAgentID error = %v", err)
	}
	if _, err := (AgentA2AClient{}).CallAgent(context.Background(), "run-parent", "target-agent", JSONMap{}, CallAgentOptions{}); err == nil {
		t.Fatalf("missing API base should fail")
	}

	var received map[string]any
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/custom/a2a/call" {
			t.Fatalf("path = %s", r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&received); err != nil {
			t.Fatal(err)
		}
		writeJSON(w, http.StatusBadGateway, JSONMap{"error": JSONMap{"code": "A2A_DOWNSTREAM", "message": "downstream failed"}})
	}))
	defer server.Close()

	client = AgentA2AClient{APIBase: server.URL, RuntimeToken: "ol_live_test"}
	_, err := client.CallAgent(context.Background(), "run-parent", "target-agent", JSONMap{"q": "hello"}, CallAgentOptions{
		Endpoint: "/custom/a2a/call",
		Reason:   "delegate",
		Metadata: JSONMap{"trace": "edge"},
	})
	if err == nil || !strings.Contains(err.Error(), "A2A_DOWNSTREAM") {
		t.Fatalf("expected downstream error, got %v", err)
	}
	if received["reason"] != "delegate" || received["metadata"].(map[string]any)["trace"] != "edge" {
		t.Fatalf("request body = %#v", received)
	}

	if _, err := runResponseToJSONMap(&openlinker.RunResponse{RunID: "run-bad", Output: make(chan int)}); err == nil {
		t.Fatalf("unmarshalable run response should fail")
	}
}
